### PR TITLE
net: websockets: Fix websocket close procedure

### LINKED ISSUE #53942: Websocket: No close message on websocket close
**Describe the bug**
When a websocket is being closed, it is expected to send a message of WEBSOCKET_FLAG_CLOSE. Although this implementations just closes the TCP connection without sending a websocket message with the flag WEBSOCKET_FLAG_CLOSE. The result is that other implementations observe an unexpected websocket close error.

**To Reproduce**
1. Build an websocket application
2. Build up a connection to a websocket server
3. Do some data transfer
4. Call a close on the websocket
5. See the server create an exception of an unexpected websocket close

**Expected behavior**
The websocket is properly closed, no error by the websocket server.

### PR DESCRIPTION
**Pull request by rlubos**
The websocket implementation did not comply with the RFC 6455 when it
comes to connection close. The websocket should send in such case Close
control frame, and wait for the reply, before closing the underlying TCP
connection. This commit fixes this behaviour.

Additionally, fix a small mutext-related bug with websockets.

Fixes #53942

### CODE REVIEW

File: subsys/net/lib/websocket/websocket.c [hunk 1]
```diff
@@ -419,6 +423,52 @@ static int websocket_interal_disconnect(struct websocket_context *ctx)
...
+	if (ret < 0) {
+		NET_ERR("[%p] Failed to send close message (err %d).", ctx, ret);
+	}
+
+	fds.fd = ctx->real_sock;
+	fds.events = ZSOCK_POLLIN;
+
+	/* Arbitrary timeout value for waiting for close reply. */
+	start = k_uptime_get();
+	timeout = 2000;
```
Comment (jukkar) [inline subsys/net/lib/websocket/websocket.c:437]: Perhaps document this what the value means

**Review by jukkar — COMMENTED** [review]
LGTM, small nit about the bare value

File: subsys/net/lib/websocket/websocket.c [hunk 2]
```diff
@@ -419,6 +423,52 @@ static int websocket_interal_disconnect(struct websocket_context *ctx)
...
+	while (timeout > 0) {
```
Comment (jukkar) [inline subsys/net/lib/websocket/websocket.c:438]: BTW, if the peer does not send a reply to the close, then we have a timeout. Just wondering if this is desired or not, should we just ignore the reply in order not to get into a timeout case, wdyt?

**Review by ssharks — COMMENTED** [review]
Thanks for sending in this PR, it seems to be pretty complete. I'll definitely give it a spin soon and provide you feedback.
I have been thinking about a similar fix, this seems to handle the case when the Zephyr side initiates the close. What happens if the other side initiates the close?
Does it make sense to create a testcase for this condition? 
I took a quick look on the existing websocket test. I do not understand how the sockets are created @jukkar. Do you know how this should work?

In the situation I use, the Zephyr side will initiate the close. Did you test the other way?

**Comment by jukkar** [comment]
> I took a quick look on the existing websocket test. I do not understand how the sockets are created @jukkar. Do you know how this should work?

The websocket tests are implemented "slightly" hackish way. It would have required lot ot extra code to implement the websocket server for the unit tests (we have the websocket client support in zephyr). Because of this, we have a check in `websocket_send_msg()` for `CONFIG_NET_TEST` which bypasses various checks and allows us to run the unit tests without actually needing to run the connect function and connect to a websocket server. Of course it would be best if we could have a simple websocket server implementation, patches are welcome of course. 

For a true network e2e tests for websocket, it could also be possible to utilize the net e2e network tests described here https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/net/README.txt. 
I managed to implement many of those and they are tracked in #29764, unfortunately the websocket_client sample is not one of the implemented test. Perhaps it could be a better idea to improve the net e2e tests for websocket and keep the unit test a simple one. Not sure about the current net e2e test status, but I suspect that no one has been running those network e2e tests and the tests might have been bit rotted atm.

File: subsys/net/lib/websocket/websocket.c (hunk 2 above)
Comment (rlubos) [reply subsys/net/lib/websocket/websocket.c:438]: I'm not sure I follow, do you suggest to skip the `zsock_poll()`/`websocket_recv_msg()` and then close the socket immediately after sending close message? I'm not sure if we'll be RFC compliant then, specifically:
> the client SHOULD wait for the server to
   close the connection but MAY close the connection at any time after
   sending **and receiving a Close message**

So in ideal conditions, the client should wait for the reply. Note, that if the server decides to close the underlying connection, the `zsock_poll()` will exit immediately.

File: subsys/net/lib/websocket/websocket.c [hunk 3]
```diff
@@ -419,6 +423,52 @@ static int websocket_interal_disconnect(struct websocket_context *ctx)
 
 	NET_DBG("[%p] Disconnecting", ctx);
 
+	ret = websocket_send_msg(ctx->sock, NULL, 0, WEBSOCKET_OPCODE_CLOSE,
+				 true, true, 0);
```
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/lib/websocket/websocket.c:427]: I think the send should be blocking in this case, will fix.

**Comment by rlubos** [comment]
> What happens if the other side initiates the close?

The Zephyr websocket client impelmentation does not handle control frames (hell, till recently they were not even process correctly). So it's up to the application right now to check if the close message was received and act accordingly.

I don't think that this is great, and IMO the websocket code should take care of dealing with control frames like close, ping etc. However  I don't think this is in the scope of this PR.

**Comment by rlubos** [comment]
> Does it make sense to create a testcase for this condition?

I'll look into websocket testsuite, i. e. how feasible is that.

File: subsys/net/lib/websocket/websocket.c (hunk 1 above)
Comment (rlubos) [reply subsys/net/lib/websocket/websocket.c:437]: I've moved the comment above just before setting the timeout, plus modified it a bit to be more precise.

File: subsys/net/lib/websocket/websocket.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/lib/websocket/websocket.c:438]: I was just reading your comment in #53942 where you mentioned that not many servers support the close message. So I am just pondering would that cause some issues with extra waits if in our side we expect a reply.

**Comment by rlubos** [comment]
> > Does it make sense to create a testcase for this condition?
> 
> I'll look into websocket testsuite, i. e. how feasible is that.

I think I'll pass on the tests for now, I don't see an easy way to incorporate such test into the current testing model (I might be missing something of course).

File: subsys/net/lib/websocket/websocket.c (hunk 2 above)
Comment (rlubos) [reply subsys/net/lib/websocket/websocket.c:438]: Well, if the server is completely out-of-spec, we'd just waste 2 seconds. But those servers that I mentioned, although did not send WebSocket close response, closed the underlying TCP connection immediately, so we're not waiting on our side.

**Comment by ssharks** [comment]
> > > Does it make sense to create a testcase for this condition?
> > 
> > 
> > I'll look into websocket testsuite, i. e. how feasible is that.
> 
> I think I'll pass on the tests for now, I don't see an easy way to incorporate such test into the current testing model (I might be missing something of course).

I agree, it is very complicated to get a test case up

**Comment by ssharks** [comment]
> > What happens if the other side initiates the close?
> 
> The Zephyr websocket client impelmentation does not handle control frames (hell, till recently they were not even process correctly). So it's up to the application right now to check if the close message was received and act accordingly.
> 
> I don't think that this is great, and IMO the websocket code should take care of dealing with control frames like close, ping etc. However I don't think this is in the scope of this PR.

For the client to be able to handle this situation correctly, upon reception of a close event, it must be able to send a close reply. Next it must also be able to close the websocket without sending the close message.
The suggested implementation does not offer this option.

**Comment by rlubos** [comment]
> > > What happens if the other side initiates the close?
> > 
> > 
> > The Zephyr websocket client impelmentation does not handle control frames (hell, till recently they were not even process correctly). So it's up to the application right now to check if the close message was received and act accordingly.
> > I don't think that this is great, and IMO the websocket code should take care of dealing with control frames like close, ping etc. However I don't think this is in the scope of this PR.
> 
> For the client to be able to handle this situation correctly, upon reception of a close event, it must be able to send a close reply. Next it must also be able to close the websocket without sending the close message. The suggested implementation does not offer this option.

Why not? Upon receiving the close control frame, the client would simply close the websocket "socket" (or call `websocket_disconnect()`). The websocket code will send a close message on the client behalf, and wait for the server to close the underlying connection. According to RFC, it's server responsibility to close the TCP connection:
> The server MUST close the underlying TCP
   connection immediately; the client SHOULD wait for the server to
   close the connection but MAY close the connection at any time after
   sending and receiving a Close message, e.g., if it has not received a
   TCP Close from the server in a reasonable time period.

Note that there's no differentiation beween close request/response, it's just a close control frame, having the same format in both cases.

**Comment by ssharks** [comment]
> > > > What happens if the other side initiates the close?
> > > 
> > > 
> > > The Zephyr websocket client impelmentation does not handle control frames (hell, till recently they were not even process correctly). So it's up to the application right now to check if the close message was received and act accordingly.
> > > I don't think that this is great, and IMO the websocket code should take care of dealing with control frames like close, ping etc. However I don't think this is in the scope of this PR.
> > 
> > 
> > For the client to be able to handle this situation correctly, upon reception of a close event, it must be able to send a close reply. Next it must also be able to close the websocket without sending the close message. The suggested implementation does not offer this option.
> 
> Why not? Upon receiving the close control frame, the client would simply close the websocket "socket" (or call `websocket_disconnect()`). The websocket code will send a close message on the client behalf, and wait for the server to close the underlying connection. According to RFC, it's server responsibility to close the TCP connection:
> 
> > The server MUST close the underlying TCP
> > connection immediately; the client SHOULD wait for the server to
> > close the connection but MAY close the connection at any time after
> > sending and receiving a Close message, e.g., if it has not received a
> > TCP Close from the server in a reasonable time period.
> 
> Note that there's no differentiation beween close request/response, it's just a close control frame, having the same format in both cases.

Maybe you are right, in that case the call.
ret = websocket_recv_msg(ctx->sock, &buf, sizeof(buf),
						 &message_type, NULL, 0);

Will return because the connection had been closed. Sound like a workable solution.

**Comment by jukkar** [comment]
> and wait for the server to close the underlying connection

Yes, and this is what I am worried about. As the connection is basically closed now, why would we need to wait the server to send us anything after the connection is closed. The websocket lib has basically worked like that already, only thing missing was the sending of the close message. From server point of view it has received the close message and should be happy now. You mentioned in the issue that many of the servers do not close the connection according to the spec. So it should do no harm if we are a bit faster with the connection closing. If the server sends us a close ack back, then this patch is good, but if it does not (could happen if the server has not implemented this properly), then we just have extra 2 sec timeout without any extra benefit as the socket is basically closed already for both parties.
I am proposing here that we send the close message, but then do not wait anything back but just proceed with socket closing.

**Comment by ssharks** [comment]
> > > and wait for the server to close the underlying connection
> > 
> > 
> > Yes, and this is what I am worried about. As the connection is basically closed now, why would we need to wait the server to send us anything after the connection is closed. The websocket lib has basically worked like that already, only thing missing was the sending of the close message. From server point of view it has received the close message and should be happy now. You mentioned in the issue that many of the servers do not close the connection according to the spec. So it should do no harm if we are a bit faster with the connection closing. If the server sends us a close ack back, then this patch is good, but if it does not (could happen if the server has not implemented this properly), then we just have extra 2 sec timeout without any extra benefit as the socket is basically closed already for both parties. I am proposing here that we send the close message, but then do not wait anything back but just proceed with socket closing.
> 
> Well, it all goes down to how close do we want to stick to the RFC, which says:
> 
> > the client SHOULD wait for the server to close the connection
> 
> Given that it is not a hard requirement (SHOULD) I think we could close TCP the connection right away as well. @ssharks Any opinion on this, shall we skip the `poll()/`recv()` part? Perhaps you could try it out with your server as it seems to be more wary about the teardown procedure?

Personally I agree with @jukkar and think we should indeed skip the poll/recv part. This will also make it possible to properly handle the situation when the server sends us a close message, without an unnecessary wait cycle.

**Comment by ssharks** [comment]
Sorry, due to activities I have not had the time to test your change yet.

**Comment by rlubos** [comment]
Ok, I've removed the `poll()`/`recvfrom()`, now the websocket will only send the Close control frame before closing the TCP connecion.

**Review by jukkar — APPROVED** [review]

**Review by ssharks — APPROVED** [review]

File: subsys/net/lib/websocket/websocket.c [hunk 4]
```diff
@@ -419,6 +419,12 @@ static int websocket_interal_disconnect(struct websocket_context *ctx)
 
 	NET_DBG("[%p] Disconnecting", ctx);
 
+	ret = websocket_send_msg(ctx->sock, NULL, 0, WEBSOCKET_OPCODE_CLOSE,
+				 true, true, SYS_FOREVER_MS);
+	if (ret < 0) {
+		NET_ERR("[%p] Failed to send close message (err %d).", ctx, ret);
+	}
+
 	ret = close(ctx->real_sock);
```
Comment (ssharks) [inline subsys/net/lib/websocket/websocket.c:428]: The return code is overwritten by the close here, is this intended?

Comment (rlubos) [reply subsys/net/lib/websocket/websocket.c:428]: Yes, IMO `close()` is essential here when it comes to the error code returned, as send could fail in certain cases (for example if the peer has already closed the TCP connection).

**Comment by ssharks** [comment]
Just validated the fix on HW and it works as far as I can tell

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**